test/cbmc/proofs/parseJSONbyModel/parseJSONbyModel_harness.c

```diff
@@ -0,0 +1,126 @@
+
+/* Stub to validate the JSON string. */
+DocParseErr_t validateJSON( const char * pJson,
+                            uint32_t messageLength )
+{
+    DocParseErr_t parseErr;
+
+    __CPROVER_assume( ( parseErr >= DocParseErrUnknown ) && ( parseErr <= DocParseErrInvalidToken ) );
+
+    return parseErr;
+}
+
+/* Stub to search for constants in a JSON string. */
+JSONStatus_t JSON_SearchConst( const char * buf,
+                               size_t max,
+                               const char * query,
+                               size_t queryLength,
+                               const char ** outValue,
+                               size_t * outValueLength,
+                               JSONTypes_t * outType )
+{
+    JSONStatus_t status;
+    size_t value;
+
+    /* valueLength cannot exceed the length of buffer.*/
+    __CPROVER_assume( value > 2u && value < OTA_FILE_BLOCK_SIZE );
+
+    *outValueLength = value;
+    *outValue = buf;
+
+    return status;
+}
+
+/* Stub to extract parameters from a JSON string. */
+DocParseErr_t extractParameter( JsonDocParam_t docParam,
+                                void * pContextBase,
+                                const char * pValueInJson,
+                                size_t valueLength )
+{
+    DocParseErr_t parseErr;
+
+    /* pContextBase and pValueInJson are defined in parseJSONbyModel function
+     * before calling extractParameter.*/
+    __CPROVER_assert( pContextBase != NULL, "Error: Expected a non-NULL pContextBase value. " );
+    __CPROVER_assert( pValueInJson != NULL, "Error: Expected a non-NULL pValueInJson value. " );
+
+    return parseErr;
+}
+
+/* Stub to Check if all the required parameters for job document are extracted from the JSON. */
+DocParseErr_t verifyRequiredParamsExtracted( const JsonDocParam_t * pModelParam,
+                                             const JsonDocModel_t * pDocModel )
+{
+    DocParseErr_t parseErr;
+
+    /* pModelParam and pDocModel are defined in parseJSONbyModel function
+     * before calling extractParameter.*/
+    __CPROVER_assert( pModelParam != NULL, "Error: Expected a non-NULL pContextBase value. " );
+    __CPROVER_assert( pDocModel != NULL, "Error: Expected a non-NULL pValueInJson value. " );
+
+    return parseErr;
+}
+
+void parseJSONbyModel_harness()
+{
+    /* pJson is always passed as a global buffer from OtaEventData_t which is
+     * enforced in processJobHandler. */
+    JsonDocModel_t docModel;
+    char pJson[ OTA_DATA_BLOCK_SIZE ];
+    uint32_t messageLength;
+
+    /* Preconditions. */
+
+    /* Havoc otaAgent to non-deterministically set all the bytes in
+     * the structure. */
+    __CPROVER_havoc_object( &otaAgent );
+
+    /* Length of the message cannot exceed the size of the message buffer. */
+    __CPROVER_assume( messageLength < OTA_DATA_BLOCK_SIZE );
+
+    /* Initialize the docModel. */
+    docModel.pBodyDef = otaJobDocModelParamStructure;
```

Comment:
These constraints seem too restrictive to properly test this function.

In particular, if you were to havoc the otaJobDocModelParamStructure object, this proof would fail for a variety of reasons.